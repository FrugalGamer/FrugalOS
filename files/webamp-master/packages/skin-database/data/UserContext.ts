// Currently only used as a WeakMap key
export default class UserContext {
  username: string | null;
  constructor(username?: string) {
    this.username = username || null;
  }
  authed() {
    return this.username != null;
  }
}

export function ctxWeakMapMemoize<T>(factory: () => T) {
  const cache: WeakMap<UserContext, T> = new WeakMap();
  return function get(ctx: UserContext): T {
    if (!cache.has(ctx)) {
      cache.set(ctx, factory());
    }
    // @ts-ignore We just put the value in there
    return cache.get(ctx);
  };
}
