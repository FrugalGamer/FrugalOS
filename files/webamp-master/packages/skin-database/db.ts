import Knex from "knex";
import * as knexConfigs from "./knexfile";
import { NODE_ENV } from "./config";

export const knex = Knex(knexConfigs[NODE_ENV ?? "test"]);
