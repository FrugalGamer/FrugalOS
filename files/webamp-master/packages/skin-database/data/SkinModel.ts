import {
  TweetStatus,
  SkinRow,
  ReviewRow,
  UploadStatus,
  SkinType,
} from "../types";
import UserContext, { ctxWeakMapMemoize } from "./UserContext";
import TweetModel, { TweetDebugData } from "./TweetModel";
import IaItemModel, { IaItemDebugData } from "./IaItemModel";
import FileModel, { FileDebugData } from "./FileModel";
import { MD5_REGEX, withUrlAsTempFile } from "../utils";
import DataLoader from "dataloader";
import { knex } from "../db";
import UploadModel, { UploadDebugData } from "./UploadModel";
import ArchiveFileModel, { ArchiveFileDebugData } from "./ArchiveFileModel";
import * as Skins from "./skins";
import fetch from "node-fetch";
import JSZip from "jszip";
import fs from "fs/promises";
import path from "path";
import { getTransparentAreaSize } from "../transparency";

export const IS_README = /(file_id\.diz)|(\.txt)$/i;
// Skinning Updates.txt ?
export const IS_NOT_README =
  /(dialogs\.txt)|(genex\.txt)|(genexinfo\.txt)|(gen_gslyrics\.txt)|(region\.txt)|(pledit\.txt)|(viscolor\.txt)|(winampmb\.txt)|("gen_ex help\.txt)|(mbinner\.txt)$/i;

export default class SkinModel {
  constructor(readonly ctx: UserContext, readonly row: SkinRow) {}

  static async fromMd5(
    ctx: UserContext,
    md5: string
  ): Promise<SkinModel | null> {
    const row = await getSkinLoader(ctx).load(md5);
    return row == null ? null : new SkinModel(ctx, row);
  }

  static clearMd5(ctx: UserContext, md5: string): void {
    getSkinLoader(ctx).clear(md5);
  }

  static async fromMd5Assert(
    ctx: UserContext,
    md5: string
  ): Promise<SkinModel> {
    const skin = await SkinModel.fromMd5(ctx, md5);
    if (skin == null) {
      throw new Error(`Expected to find skin with md5 "${md5}".`);
    }
    return skin;
  }

  static async fromAnything(
    ctx: UserContext,
    anything: string
  ): Promise<SkinModel | null> {
    const md5Match = anything.match(MD5_REGEX);
    if (md5Match != null) {
      const md5 = md5Match[1];
      const found = await SkinModel.fromMd5(ctx, md5);
      if (found != null) {
        return found;
      }
    }
    const iaItem = await IaItemModel.fromAnything(ctx, anything);
    if (iaItem != null) {
      return iaItem.getSkin();
    }
    const tweet = await TweetModel.fromAnything(ctx, anything);
    if (tweet != null) {
      return tweet.getSkin();
    }
    return null;
  }

  static async exists(ctx: UserContext, md5: string): Promise<boolean> {
    const row = await getSkinLoader(ctx).load(md5);
    return row != null;
  }

  getSkinType(): SkinType {
    switch (this.row.skin_type) {
      case 1:
        return "CLASSIC";
      case 2:
        return "MODERN";
    }
    throw new Error(`Unknown skin_type ${this.row.skin_type}`);
  }

  getId(): number {
    return this.row.id;
  }

  async tweeted(): Promise<boolean> {
    return (await this.getTweet()) != null;
  }

  async getTweet(): Promise<TweetModel | null> {
    const tweets = await this.getTweets();
    return tweets[0] || null;
  }

  async getTweets(): Promise<TweetModel[]> {
    return TweetModel.fromMd5(this.ctx, this.row.md5);
  }

  getIaItem(): Promise<IaItemModel | null> {
    return IaItemModel.fromMd5(this.ctx, this.row.md5);
  }

  getReviews(): Promise<ReviewRow[]> {
    return getReviewsLoader(this.ctx).load(this.row.md5);
  }

  getFiles(): Promise<FileModel[]> {
    return FileModel.fromMd5(this.ctx, this.row.md5);
  }

  getArchiveFiles(): Promise<ArchiveFileModel[]> {
    return ArchiveFileModel.fromMd5(this.ctx, this.row.md5);
  }

  getUploadStatuses(): Promise<UploadModel[]> {
    return UploadModel.fromMd5(this.ctx, this.row.md5);
  }

  async getUploadStatus(): Promise<UploadStatus | null> {
    const status = await Skins.getUploadStatuses([this.row.md5]);
    return (status[this.getMd5()] as UploadStatus) || null;
  }

  async getIsNsfw(): Promise<boolean> {
    const reviews = await this.getReviews();
    return reviews.some((review) => review.review === "NSFW");
  }

  async getTweetStatus(): Promise<TweetStatus> {
    const tweet = await this.getTweet();
    if (tweet != null) {
      return "TWEETED";
    }
    const reviewRows = await this.getReviews();
    const reviews = new Set(reviewRows.map((row) => row.review));
    if (reviews.has("NSFW")) {
      return "NSFW";
    }
    if (reviews.has("REJECTED")) {
      return "REJECTED";
    }
    if (reviews.has("APPROVED")) {
      return "APPROVED";
    }

    return "UNREVIEWED";
  }

  async getFileName(): Promise<string> {
    const files = await this.getFiles();
    if (files.length === 0) {
      throw new Error(`Could not find file for skin with md5 ${this.getMd5()}`);
    }
    const filename = files[0].getFileName();
    if (!filename.match(/\.(zip)|(wsz)|(wal)$/i)) {
      throw new Error("Expected filename to end with zip, wsz or wal.");
    }
    return filename;
  }

  async getScreenshotFileName(): Promise<string> {
    if (this.getSkinType() === "MODERN") {
      throw new Error("Modern skins do not have screenshots yet.");
    }
    const skinFilename = await this.getFileName();
    return skinFilename.replace(/\.(wsz|zip)$/, ".png");
  }

  getMd5(): string {
    return this.row.md5;
  }

  getEmails(): string[] {
    const { emails } = this.row;
    return emails ? emails.split(" ") : [];
  }

  async getReadme(): Promise<string | null> {
    const files = await this.getArchiveFiles();
    const readme = files.find((file) => {
      const filename = file.getFileName();
      return IS_README.test(filename) && !IS_NOT_README.test(filename);
    });

    if (readme == null) {
      return null;
    }
    return readme.getTextContent();
  }

  getMuseumUrl(): string {
    if (this.getSkinType() === "MODERN") {
      throw new Error("Modern skins do not render in the museum.");
    }
    return `https://skins.webamp.org/skin/${this.row.md5}`;
  }
  getWebampUrl(): string {
    if (this.getSkinType() === "MODERN") {
      throw new Error("Modern skins do not render in Webamp.");
    }
    return `https://webamp.org?skinUrl=${this.getSkinUrl()}`;
  }
  getScreenshotUrl(): string {
    if (this.getSkinType() === "MODERN") {
      throw new Error("Modern skins do not have screenshots.");
    }
    return Skins.getScreenshotUrl(this.row.md5);
  }
  getSkinUrl(): string {
    switch (this.getSkinType()) {
      case "CLASSIC":
        return Skins.getSkinUrl(this.row.md5);
      case "MODERN":
        return Skins.getModernSkinUrl(this.row.md5);
    }
  }

  getAverageColor(): string {
    return this.row.average_color;
  }

  getBuffer = mem(async (): Promise<Buffer> => {
    if (process.env.LOCAL_FILE_CACHE) {
      const ext = this.getSkinType() === "CLASSIC" ? ".wsz" : ".wal";
      const skinPath = path.join(
        process.env.LOCAL_FILE_CACHE,
        "skins",
        this.getMd5() + ext
      );
      return fs.readFile(skinPath);
    } else {
      const response = await fetch(this.getSkinUrl());
      if (!response.ok) {
        throw new Error(`Could not fetch skin at "${this.getSkinUrl()}"`);
      }
      return response.buffer();
    }
  });

  getScreenshotBuffer = mem(async (): Promise<Buffer> => {
    if (this.getSkinType() === "MODERN") {
      throw new Error("Modern skins do not have screenshots.");
    }
    const response = await fetch(this.getScreenshotUrl());
    if (!response.ok) {
      throw new Error(
        `Could not fetch skin screenshot at "${this.getScreenshotUrl()}"`
      );
    }
    return response.buffer();
  });

  async withTempFile(cb: (file: string) => Promise<void>): Promise<void> {
    const filename = await this.getFileName();

    return withUrlAsTempFile(this.getSkinUrl(), filename, cb);
  }

  async _hasSpriteSheet(base: string): Promise<boolean> {
    const ext = "(bmp)|(png)";
    return this._hasFile(base, ext);
  }

  async _hasFile(base: string, ext: string): Promise<boolean> {
    // TODO: Pre-compile regexp
    const matcher = new RegExp(`^(.*[/\\\\])?${base}.(${ext})$`, "i");
    const archiveFiles = await this.getArchiveFiles();
    return archiveFiles.some((file) => {
      return matcher.test(file.getFileName());
    });
  }

  async _getFile(base: string, ext: string): Promise<ArchiveFileModel | null> {
    // TODO: Pre-compile regexp
    const matcher = new RegExp(`^(.*[/\\\\])?${base}.(${ext})$`, "i");
    const archiveFiles = await this.getArchiveFiles();
    const row = archiveFiles.find((file) => {
      return matcher.test(file.getFileName());
    });
    return row || null;
  }

  async hasEqualizer(): Promise<boolean> {
    return this._hasSpriteSheet("EQMAIN");
  }
  async hasPlaylist(): Promise<boolean> {
    return this._hasSpriteSheet("PLEDIT");
  }
  async hasMediaLibrary(): Promise<boolean> {
    return this.hasGeneral();
  }

  async hasBrowser(): Promise<boolean> {
    return this._hasSpriteSheet("MB");
  }

  async hasAVS(): Promise<boolean> {
    return this._hasSpriteSheet("AVS");
  }

  async hasVideo(): Promise<boolean> {
    return this._hasSpriteSheet("VIDEO");
  }

  // Has built-in support for the MikroAMP plugin.
  async hasMikro(): Promise<boolean> {
    // Could also check for `WINAMPMB.TXT`.
    return this._hasSpriteSheet("MIKRO");
  }

  // Has built-in support of the Amarok plugin.
  async hasAmarok(): Promise<boolean> {
    return this._hasSpriteSheet("AMAROK");
  }

  // Has built-in support of the vidamp
  async hasVidamp(): Promise<boolean> {
    return this._hasSpriteSheet("VIDAMP");
  }

  // Includes custom cursors
  async hasCur(): Promise<boolean> {
    const matcher = new RegExp(`.(cur)$`, "i");
    const archiveFiles = await this.getArchiveFiles();
    return archiveFiles.some((file) => {
      return matcher.test(file.getFileName());
    });
  }

  // Has transparency
  async hasTransparency(): Promise<boolean> {
    const size = await this.transparentAreaSize();
    return size > 0;
  }

  async transparentPixels(): Promise<number> {
    const region = await this._getFile("region", "txt");
    if (region == null) {
      return 0;
    }
    const text = await region.getTextContent();
    if (text == null) {
      return 0;
    }
    try {
      return getTransparentAreaSize(text);
    } catch (e) {
      console.error(`Failed: ${this.getMd5()}`);
      return 0;
    }
  }

  async hasAni(): Promise<boolean> {
    // Note: This should be expanded to check for animated cursors that use the
    // .cur extension (but are actually .ani under the hood).
    const matcher = new RegExp(`.(ani)$`, "i");
    const archiveFiles = await this.getArchiveFiles();
    return archiveFiles.some((file) => {
      return matcher.test(file.getFileName());
    });
  }

  async hasGeneral(): Promise<boolean> {
    return (
      (await this._hasSpriteSheet("GEN")) &&
      (await this._hasSpriteSheet("GENEX"))
    );
  }

  async getAlgoliaIndexUpdates(limit?: number): Promise<any[]> {
    return Skins.searchIndexUpdatesForSkin(this.getMd5(), limit);
  }

  async withScreenshotTempFile<T>(
    cb: (file: string) => Promise<T>
  ): Promise<T> {
    if (this.getSkinType() === "MODERN") {
      throw new Error("Modern skins do not have screenshots.");
    }
    const screenshotFilename = await this.getScreenshotFileName();
    return withUrlAsTempFile(this.getScreenshotUrl(), screenshotFilename, cb);
  }

  getZip = mem(async (): Promise<JSZip> => {
    const buffer = await this.getBuffer();
    return JSZip.loadAsync(buffer);
  });

  async debug(): Promise<{
    row: SkinRow;
    reviews: ReviewRow[];
    tweets: TweetDebugData[];
    files: FileDebugData[];
    archiveFiles: ArchiveFileDebugData[];
    iaItem: IaItemDebugData | null;
    uploadStatuses: UploadDebugData[];
  }> {
    const tweets = await this.getTweets();
    const files = await this.getFiles();
    const archiveFiles = await this.getArchiveFiles();
    const iaItem = await this.getIaItem();
    const uploadStatuses = await this.getUploadStatuses();
    return {
      row: this.row,
      reviews: await this.getReviews(),
      tweets: await Promise.all(tweets.map((tweet) => tweet.debug())),
      files: await Promise.all(files.map((file) => file.debug())),
      archiveFiles: await Promise.all(archiveFiles.map((file) => file.debug())),
      uploadStatuses: await Promise.all(
        uploadStatuses.map((upload) => upload.debug())
      ),
      iaItem: iaItem == null ? null : await iaItem.debug(),
    };
  }
}

const getSkinLoader = ctxWeakMapMemoize<DataLoader<string, SkinRow | null>>(
  () =>
    new DataLoader(async (md5s) => {
      const rows = await knex("skins").whereIn("md5", md5s).select();
      return md5s.map((md5) => rows.find((x) => x.md5 === md5));
    })
);

const getReviewsLoader = ctxWeakMapMemoize<DataLoader<string, ReviewRow[]>>(
  () =>
    new DataLoader(async (md5s) => {
      const rows = await knex("skin_reviews")
        .whereIn("skin_md5", md5s)
        .select();
      return md5s.map((md5) => rows.filter((x) => x.skin_md5 === md5));
    })
);

function mem<T>(fn: () => T): () => T {
  let cached: T | null = null;
  return () => {
    if (cached == null) {
      cached = fn();
    }
    return cached;
  };
}
