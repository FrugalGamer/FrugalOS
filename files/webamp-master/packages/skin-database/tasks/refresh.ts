import UserContext from "../data/UserContext";
import SkinModel from "../data/SkinModel";
import { knex } from "../db";
import { setHashesForSkin } from "../skinHash";
import Shooter from "../shooter";
import { screenshot } from "./screenshotSkin";
import * as Skins from "../data/skins";
import { SkinType } from "../types";
import { getSkinType } from "../addSkin";

// TODO Move this into the function so that we clean up on each run?

async function getExtractionError(skin: SkinModel): Promise<string | null> {
  try {
    await skin.getZip();
  } catch (e) {
    return e.message;
  }

  return null;
}

// A task to download a skin and refresh all the data we scrape from it

export async function getSkinsToRefresh(
  ctx: UserContext,
  n: number,
  neverDone = false
): Promise<SkinModel[]> {
  let query = knex("skins")
    .leftJoin("refreshes", "skins.md5", "refreshes.skin_md5")
    .where("skin_type", 1)
    .orderBy("refreshes.timestamp", "asc")
    .limit(n);

  if (neverDone) {
    query = query.where("refreshes.skin_md5", null);
  }
  const skins = await query.select();

  return skins.map((row) => new SkinModel(ctx, row));
}

export async function refreshSkins(
  skins: SkinModel[],
  options: { noScreenshot: boolean }
): Promise<void> {
  if (options.noScreenshot) {
    while (skins.length > 0) {
      const skin = skins.pop();
      // @ts-ignore
      await refresh(skin, null);
    }
    return;
  }
  const shooterLogger = () => {
    // Don't log
  };
  await Shooter.withShooter(async (shooter: Shooter) => {
    for (const [i, skin] of skins.entries()) {
      await refresh(skin, shooter);
      // We end up caching a lot of stuff (the whole skin/zip) on the model, so we can't just leave these around for the whole process.
      delete skins[i];
    }
  }, shooterLogger);
}

export async function _refresh(
  skin: SkinModel,
  shooter: Shooter | null
): Promise<void> {
  const extractionError = await getExtractionError(skin);
  if (extractionError != null) {
    throw new Error(extractionError);
  }
  await setHashesForSkin(skin);
  await Skins.setContentHash(skin.getMd5());

  let skinType: SkinType;
  try {
    skinType = await getSkinType(await skin.getZip());
  } catch (e) {
    throw new Error("Not a skin (no main.bmp/skin.xml)");
  }
  if (skinType !== "CLASSIC") {
    // TODO:
    // Reviews are invalid
    // Delete from algolia
    // Delete screenshot
    // Tweets are invalid
    throw new Error("Skin type mismatch");
  }

  // Retake screenshot
  if (shooter != null) {
    await screenshot(skin, shooter);
  }

  await knex("refreshes").insert({
    skin_md5: skin.getMd5(),
  });
}

export async function refresh(
  skin: SkinModel,
  shooter: Shooter | null
): Promise<void> {
  if (skin.getSkinType() !== "CLASSIC") {
    throw new Error("Can't refresh non-classic skins");
  }
  try {
    await _refresh(skin, shooter);
  } catch (e) {
    console.error(e);
    await knex("refreshes").insert({
      skin_md5: skin.getMd5(),
      error: e.message,
    });
    return;
  }
}
