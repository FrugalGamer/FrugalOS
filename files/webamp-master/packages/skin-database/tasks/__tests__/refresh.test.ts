import path from "path";
import UserContext from "../../data/UserContext";
import { knex } from "../../db";
import { getSkinsToRefresh, refresh } from "../refresh";
import SkinModel from "../../data/SkinModel";
import fs from "fs";
import Shooter from "../../shooter";

jest.mock("../../algolia");
jest.mock("../../s3");
jest.mock("../screenshotSkin");
jest.mock("node-fetch", () => jest.fn());

// @ts-ignore
const shooter: Shooter = {};

beforeEach(async () => {
  await knex.migrate.latest();
  await knex.seed.run();
});

test("refresh", async () => {
  const ctx = new UserContext();
  const [skin] = await getSkinsToRefresh(ctx, 1);
  expect(skin.getMd5()).toEqual("a_fake_md5");

  skin.getBuffer = async () => Buffer.from("");

  const originalConsoleError = console.error;
  console.error = jest.fn();
  await refresh(skin, shooter);
  console.error = originalConsoleError;

  const [nextToRefresh] = await getSkinsToRefresh(ctx, 1);
  expect(nextToRefresh.getMd5()).not.toEqual("a_fake_md5");
});

test("can't extract", async () => {
  const originalConsoleError = console.error;
  console.error = jest.fn();
  const ctx = new UserContext();
  const skin = await SkinModel.fromMd5Assert(ctx, "a_fake_md5");
  skin.getBuffer = async () => Buffer.from("");

  await refresh(skin, shooter);
  const refreshRow = await knex("refreshes")
    .where("skin_md5", skin.getMd5())
    .first();
  expect(refreshRow).toEqual({
    error:
      "End of data reached (data length = 0, asked index = 4). Corrupted zip ?",
    skin_md5: "a_fake_md5",
    id: expect.any(Number),
    timestamp: expect.stringMatching(/^[0-9]{4}-/),
  });
  console.error = originalConsoleError;
});

// TODO: Turn back on once dates are stable in output
test.skip("valid skin (TopazAmp)", async () => {
  const ctx = new UserContext();
  const skin = await SkinModel.fromMd5Assert(ctx, "a_fake_md5");

  skin.getBuffer = async () => {
    return fs.readFileSync(
      path.join(__dirname, "../../../webamp/demo/skins/TopazAmp1-2.wsz")
    );
  };

  await refresh(skin, shooter);

  // Check that it set the archive files
  const archiveFiles = await skin.getArchiveFiles();

  // Check archive file
  expect(
    await Promise.all(archiveFiles.map((file) => file.debug()))
  ).toMatchSnapshot();

  // Check readme
  const skinRow = await knex("skins").where("md5", skin.getMd5()).first();
  expect(skinRow.readme_text).toMatchSnapshot();
  expect(skinRow.skin_type).toBe(1);

  // Check Refresh
  const refreshRow = await knex("refreshes")
    .where("skin_md5", skin.getMd5())
    .first();
  expect(refreshRow).toEqual({
    error: null,
    skin_md5: "a_fake_md5",
    id: expect.any(Number),
    timestamp: expect.stringMatching(/^[0-9]{4}-/),
  });
});
