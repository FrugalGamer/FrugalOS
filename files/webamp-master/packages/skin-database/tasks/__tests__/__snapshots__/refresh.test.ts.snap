// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`valid skin (TopazAmp) 1`] = `
Array [
  Object {
    "row": Object {
      "file_date": "Tue Jul 20 1999 19:23:28 GMT-0700 (Pacific Daylight Time)",
      "file_md5": "fbee37aa8e23945860c1e58cd6b8a80c",
      "file_name": "Balance.bmp",
      "id": 7,
      "is_directory": 0,
      "skin_md5": "a_fake_md5",
      "text_content": null,
      "uncompressed_size": null,
    },
  },
  Object {
    "row": Object {
      "file_date": "Sat Jul 17 1999 14:14:00 GMT-0700 (Pacific Daylight Time)",
      "file_md5": "c6005d7d77f3ff78b57749f569582161",
      "file_name": "Cbuttons.bmp",
      "id": 12,
      "is_directory": 0,
      "skin_md5": "a_fake_md5",
      "text_content": null,
      "uncompressed_size": null,
    },
  },
  Object {
    "row": Object {
      "file_date": "Tue Jul 27 1999 04:15:54 GMT-0700 (Pacific Daylight Time)",
      "file_md5": "fdb3aee2834b78d87f760962d969779b",
      "file_name": "Eqmain.bmp",
      "id": 3,
      "is_directory": 0,
      "skin_md5": "a_fake_md5",
      "text_content": null,
      "uncompressed_size": null,
    },
  },
  Object {
    "row": Object {
      "file_date": "Tue Jul 27 1999 04:05:56 GMT-0700 (Pacific Daylight Time)",
      "file_md5": "c0d761d8f3363b13f99b4e691787a8eb",
      "file_name": "Main.bmp",
      "id": 5,
      "is_directory": 0,
      "skin_md5": "a_fake_md5",
      "text_content": null,
      "uncompressed_size": null,
    },
  },
  Object {
    "row": Object {
      "file_date": "Sat Jul 17 1999 14:14:18 GMT-0700 (Pacific Daylight Time)",
      "file_md5": "96040c73c8f4a0a8bda59be8f5cf0bef",
      "file_name": "Monoster.bmp",
      "id": 9,
      "is_directory": 0,
      "skin_md5": "a_fake_md5",
      "text_content": null,
      "uncompressed_size": null,
    },
  },
  Object {
    "row": Object {
      "file_date": "Sat Jul 17 1999 14:14:12 GMT-0700 (Pacific Daylight Time)",
      "file_md5": "0ad5b85bb51123c70e13b227eb689d3c",
      "file_name": "Playpaus.bmp",
      "id": 11,
      "is_directory": 0,
      "skin_md5": "a_fake_md5",
      "text_content": null,
      "uncompressed_size": null,
    },
  },
  Object {
    "row": Object {
      "file_date": "Sat Jul 24 1999 08:33:22 GMT-0700 (Pacific Daylight Time)",
      "file_md5": "44faaa170461ca54203124892d697654",
      "file_name": "Pledit.bmp",
      "id": 6,
      "is_directory": 0,
      "skin_md5": "a_fake_md5",
      "text_content": null,
      "uncompressed_size": null,
    },
  },
  Object {
    "row": Object {
      "file_date": "Sat Jul 17 1999 14:13:56 GMT-0700 (Pacific Daylight Time)",
      "file_md5": "d7e938bdec0aa7490977450d7162446e",
      "file_name": "Posbar.bmp",
      "id": 13,
      "is_directory": 0,
      "skin_md5": "a_fake_md5",
      "text_content": null,
      "uncompressed_size": null,
    },
  },
  Object {
    "row": Object {
      "file_date": "Sat Jul 17 1999 14:13:46 GMT-0700 (Pacific Daylight Time)",
      "file_md5": "9d12fcaf2bad4fdf9f75c00554a051c2",
      "file_name": "Shufrep.bmp",
      "id": 15,
      "is_directory": 0,
      "skin_md5": "a_fake_md5",
      "text_content": null,
      "uncompressed_size": null,
    },
  },
  Object {
    "row": Object {
      "file_date": "Sat Jul 17 1999 14:13:52 GMT-0700 (Pacific Daylight Time)",
      "file_md5": "7a08e36a3f7607dff03fb54b9cc95621",
      "file_name": "Text.bmp",
      "id": 14,
      "is_directory": 0,
      "skin_md5": "a_fake_md5",
      "text_content": null,
      "uncompressed_size": null,
    },
  },
  Object {
    "row": Object {
      "file_date": "Tue Jul 27 1999 04:18:58 GMT-0700 (Pacific Daylight Time)",
      "file_md5": "2d58324dece68707e973ee5f4f072f7a",
      "file_name": "Titlebar.bmp",
      "id": 2,
      "is_directory": 0,
      "skin_md5": "a_fake_md5",
      "text_content": null,
      "uncompressed_size": null,
    },
  },
  Object {
    "row": Object {
      "file_date": "Tue Jul 27 1999 04:10:20 GMT-0700 (Pacific Daylight Time)",
      "file_md5": "890df1888deb6e37ad34077d6e433a1e",
      "file_name": "eq_ex.bmp",
      "id": 4,
      "is_directory": 0,
      "skin_md5": "a_fake_md5",
      "text_content": null,
      "uncompressed_size": null,
    },
  },
  Object {
    "row": Object {
      "file_date": "Sat Jul 17 1999 14:13:28 GMT-0700 (Pacific Daylight Time)",
      "file_md5": "c3f72ef137c58ae52e825e41a2129ca8",
      "file_name": "mb.bmp",
      "id": 16,
      "is_directory": 0,
      "skin_md5": "a_fake_md5",
      "text_content": null,
      "uncompressed_size": null,
    },
  },
  Object {
    "row": Object {
      "file_date": "Sat Jul 17 1999 14:14:20 GMT-0700 (Pacific Daylight Time)",
      "file_md5": "c9ef2c8a717f354c6dccf839de4237d2",
      "file_name": "numbers.bmp",
      "id": 8,
      "is_directory": 0,
      "skin_md5": "a_fake_md5",
      "text_content": null,
      "uncompressed_size": null,
    },
  },
  Object {
    "row": Object {
      "file_date": "Sat Jul 17 1999 14:14:14 GMT-0700 (Pacific Daylight Time)",
      "file_md5": "654653f2114144a2c079f4edfd58b5e1",
      "file_name": "nums_ex.bmp",
      "id": 10,
      "is_directory": 0,
      "skin_md5": "a_fake_md5",
      "text_content": null,
      "uncompressed_size": null,
    },
  },
  Object {
    "row": Object {
      "file_date": "Sat Jul 17 1999 09:58:44 GMT-0700 (Pacific Daylight Time)",
      "file_md5": "c378b0acc505489f07adeabe9e6f3cff",
      "file_name": "pledit.txt",
      "id": 18,
      "is_directory": 0,
      "skin_md5": "a_fake_md5",
      "text_content": null,
      "uncompressed_size": null,
    },
  },
  Object {
    "row": Object {
      "file_date": "Tue Jul 27 1999 04:27:56 GMT-0700 (Pacific Daylight Time)",
      "file_md5": "baf5a61397990b5fbf5cafeadf5e5acb",
      "file_name": "readme.txt",
      "id": 1,
      "is_directory": 0,
      "skin_md5": "a_fake_md5",
      "text_content": null,
      "uncompressed_size": null,
    },
  },
  Object {
    "row": Object {
      "file_date": "Sat Jul 17 1999 09:52:36 GMT-0700 (Pacific Daylight Time)",
      "file_md5": "458b893975e52f6cb35d9d7b8d9b2bb3",
      "file_name": "viscolor.txt",
      "id": 19,
      "is_directory": 0,
      "skin_md5": "a_fake_md5",
      "text_content": null,
      "uncompressed_size": null,
    },
  },
  Object {
    "row": Object {
      "file_date": "Sat Jul 17 1999 14:13:24 GMT-0700 (Pacific Daylight Time)",
      "file_md5": "dc770d401d56e758ba64b57dcf999186",
      "file_name": "volume.bmp",
      "id": 17,
      "is_directory": 0,
      "skin_md5": "a_fake_md5",
      "text_content": null,
      "uncompressed_size": null,
    },
  },
]
`;

exports[`valid skin (TopazAmp) 2`] = `null`;
